DolbyIO/Source/DolbyIOModule/Private/DolbyIO/DolbyIO.cpp

```diff
@@ -0,0 +1,156 @@
+// Copyright 2022 Dolby Laboratories
+
+#include "DolbyIO.h"
+
+#include "DolbyIO/Authenticator.h"
+#include "DolbyIO/SdkAccess.h"
+
+#include "Async/Async.h"
+#include "Engine/GameInstance.h"
+#include "Engine/World.h"
+#include "GameFramework/PlayerController.h"
+#include "TimerManager.h"
+
+void UDolbyIO::Initialize(FSubsystemCollectionBase& Collection)
+{
+	Super::Initialize(Collection);
+	CppSdk = MakeShared<DolbyIO::FSdkAccess>(*this);
+	Authenticator = MakeShared<DolbyIO::FAuthenticator>(*this);
+	GameInstance = GetGameInstance();
+
+	OnTokenNeeded();
+
+	if (GameInstance)
+	{
+		GameInstance->GetTimerManager().SetTimer(SpatialUpdateTimerHandle, this,
+		                                         &UDolbyIO::UpdateViewPointUsingFirstPlayer, 0.03, true);
+	}
+}
+
+void UDolbyIO::Deinitialize()
+{
+	bIsAlive = false;
+	CppSdk.Reset();
+	Super::Deinitialize();
+}
+
+void UDolbyIO::SetToken(const FString& Token)
+{
+	CppSdk->SetToken(Token);
+}
+void UDolbyIO::SetTokenUsingKeyAndSecret(const FString& AppKey, const FString& AppSecret,
+                                         int TokenExpirationTimeInSeconds)
+{
+	Authenticator->GetToken(AppKey, AppSecret, TokenExpirationTimeInSeconds);
+}
+void UDolbyIO::Connect(const FString& ConferenceName, const FString& UserName)
+{
+	CppSdk->Connect(ConferenceName, UserName);
+}
+void UDolbyIO::Disconnect()
+{
+	CppSdk->Disconnect();
+}
+void UDolbyIO::UpdateViewPoint(const FVector& Position, const FRotator& Rotation)
+{
+	if (SpatialUpdateTimerHandle.IsValid())
+	{
+		GameInstance->GetTimerManager().ClearTimer(SpatialUpdateTimerHandle);
+	}
+
+	CppSdk->UpdateViewPoint(Position, Rotation);
+}
+void UDolbyIO::MuteInput()
+{
+	CppSdk->MuteInput();
+}
+void UDolbyIO::UnmuteInput()
+{
+	CppSdk->UnmuteInput();
+}
+void UDolbyIO::MuteOutput()
+{
+	CppSdk->MuteOutput();
+}
+void UDolbyIO::UnmuteOutput()
+{
+	CppSdk->UnmuteOutput();
+}
+void UDolbyIO::GetAudioLevels()
+{
+	CppSdk->GetAudioLevels();
+}
+
+void UDolbyIO::UpdateViewPointUsingFirstPlayer()
+{
+	if (GameInstance)
```

Comment:
the `else` cases should be handled with logs to make error conditions more explicit.

Reply:
To be honest I think it would just be code bloat. The checks themselves are kind of unnecessary as they should never fail (they're more asserts than ifs), but if they didn't exist and somehow these pointers were NULL, it would cause a crash.